contracts/base/LiquidityManagement.sol

```diff
@@ -58,15 +67,34 @@ abstract contract LiquidityManagement is IUniswapV3MintCallback, PeripheryImmuta
 
         pool = IUniswapV3Pool(PoolAddress.computeAddress(factory, poolKey));
 
+        // compute the liquidity amount
+        {
+            (uint160 sqrtPriceX96, , , , , , ) = pool.slot0();
+            uint160 sqrtRatioAX96 = TickMath.getSqrtRatioAtTick(params.tickLower);
+            uint160 sqrtRatioBX96 = TickMath.getSqrtRatioAtTick(params.tickUpper);
+
+            liquidity = LiquidityAmounts.getLiquidityForAmounts(
+                sqrtPriceX96,
+                sqrtRatioAX96,
+                sqrtRatioBX96,
+                params.amount0Desired,
+                params.amount1Desired
+            );
+        }
+
         (amount0, amount1) = pool.mint(
             params.recipient,
             params.tickLower,
             params.tickUpper,
-            params.amount,
-            abi.encode(MintCallbackData({poolKey: poolKey, payer: msg.sender}))
+            liquidity,
+            abi.encode(
+                MintCallbackData({
+                    poolKey: poolKey,
+                    payer: msg.sender,
+                    amount0Min: params.amount0Min,
```

Comment:
thinking is it's slightly cheaper in case of revert to do it in the callback

Reply:
i think we should optimize for the non-reverting case